File: backend/apps/core/cpf.py
import hashlib
import hmac
import re

from django.conf import settings


def normalize(cpf: str) -> str:
    return re.sub(r"\D", "", cpf)


def is_valid(cpf: str) -> bool:
    cpf = normalize(cpf)
    if len(cpf) != 11:
        return False
    if len(set(cpf)) == 1:
        return False
    for i in range(2):
        total = sum(int(cpf[j]) * (10 + i - j) for j in range(9 + i))
        expected = (total * 10 % 11) % 10
        if expected != int(cpf[9 + i]):
            return False
    return True


def hash_cpf(cpf: str) -> str:
    cpf = normalize(cpf)
    key = settings.CPF_HMAC_KEY.encode()
    return hmac.new(key, cpf.encode(), hashlib.sha256).hexdigest()


def last2(cpf: str) -> str:
    cpf = normalize(cpf)
    return cpf[-2:]


def mask(cpf: str) -> str:
    cpf = normalize(cpf)
    if len(cpf) == 11:
        return f"***.***.***.{cpf[-2:]}"
    return "***.***.***-**"

File: frontend/src/lib/cpf.ts
export function normalize(cpf: string): string {
  return cpf.replace(/\D/g, "");
}

export function isValid(cpf: string): boolean {
  const n = normalize(cpf);
  if (n.length !== 11) return false;
  if (/^(\d)\1{10}$/.test(n)) return false;

  let sum = 0;
  for (let i = 0; i < 9; i++) sum += parseInt(n[i]) * (10 - i);
  let d1 = (sum * 10) % 11;
  if (d1 === 10 || d1 === 11) d1 = 0;
  if (d1 !== parseInt(n[9])) return false;

  sum = 0;
  for (let i = 0; i < 10; i++) sum += parseInt(n[i]) * (11 - i);
  let d2 = (sum * 10) % 11;
  if (d2 === 10 || d2 === 11) d2 = 0;
  return d2 === parseInt(n[10]);
}

export function format(cpf: string): string {
  const n = normalize(cpf);
  if (n.length <= 3) return n;
  if (n.length <= 6) return `${n.slice(0, 3)}.${n.slice(3)}`;
  if (n.length <= 9) return `${n.slice(0, 3)}.${n.slice(3, 6)}.${n.slice(6)}`;
  return `${n.slice(0, 3)}.${n.slice(3, 6)}.${n.slice(6, 9)}-${n.slice(9, 11)}`;
}

export function applyMask(value: string): string {
  const n = normalize(value).slice(0, 11);
  return format(n);
}
